presto-main/src/main/java/com/facebook/presto/execution/SqlTaskExecutionFactory.java

```diff
@@ -82,7 +91,8 @@ public SqlTaskExecution create(Session session, QueryContext queryContext, TaskS
                         fragment.getPartitioningScheme(),
                         fragment.getStageExecutionDescriptor(),
                         fragment.getPartitionedSources(),
-                        outputBuffer);
+                        outputBuffer,
+                        exchangeClientManager);
```

Comment:
I wonder if the `SqlTaskExecution` itself has to have the `removeRemoteSource` method, so the `ExchangeClientManager` can be created even within the `SqlTaskExecution`. It would simplify the interfaces, as you can directly inject the `ExchangeClientSupplier` into the `SqlTaskExecutionFactory`, and keep track of opened exchange clients withn the `SqlTaskExecution`

Reply:
@arhimondr : So get the `SqlTaskExecution` from the `taskHolderReference`  ? :)

Reply:
@arhimondr : Per offline discussion, we should do it -- I didn't do it at beginning since `TaskHolder` looks confusing to me. But after we looked into the code, there is 1:1 relationship between `SqlTask` and `SqlTaskExecution`.

Reply:
Talked offline -- `SqlTaskExecution` only work on execution related stuff -- receive splits update, creates new drivers, etc, while `SqlTask` does more "administrative" things such as abort tasks, clean output buffer. So removing remote source sounds the right responsibility  for `SqlTask`